reconcile/utils/oc_map.py

```diff
@@ -67,24 +67,22 @@ def __init__(
         # happy with regular dedicated-admin and will request a cluster
         # with oc_map.get(cluster) without specifying privileged access
         # specifically
-        privileged_clusters: list[OCConnectionParameters] = [
-            c for c in connection_parameters if (c.is_cluster_admin or cluster_admin)
-        ]
-        unprivileged_clusters: list[OCConnectionParameters] = [
-            c
+        all_clusters = {c.cluster_name: c for c in connection_parameters}
+        privileged_clusters = {
```

Comment:
Only `.values()` are used for these 2, can we use list instead of dict here? Or are we using dict to remove duplicate values by `cluster_name` (last one wins)?

Reply:
Good observation :+1: 

Yeah by using dicts Im trying to reduce the thread overhead.

The current behavior is random one wins with extra threading overhead, see [here](https://github.com/app-sre/qontract-reconcile/blob/ee0f52bf7f2551d00833d6858861ba16d1dd2357/reconcile/utils/oc_map.py#L106-L109).

Now, the original behavior was last one wins, see [here](https://github.com/app-sre/qontract-reconcile/blob/ee0f52bf7f2551d00833d6858861ba16d1dd2357/reconcile/utils/oc.py#L1586-L1590).

Whether last one wins or random one wins doesn't matter, but I would like to reduce the overhead of creating threads unnecessarily and align closer with behavior from original `OC_Map` class.

Reply:
Not sure if these are real cases:
1. Given there are 2 items in `connection_parameters` with same `cluster_name`, one `is_cluster_admin` True, the other one `is_cluster_admin` False
    When create `all_clusters`
    Then last one in `connection_parameters` will be the only one in `all_clusters`
    Will this cause trouble?
2. Given there are 2 items in `connection_parameters` with same `cluster_name`, both `is_cluster_admin` True
    When create `privileged_clusters`
    Then last one in `connection_parameters` will be the only one in `privileged_clusters`
    Will this cause trouble?

If deduplication is happening here, and it doesn't matter which connection parameter is picked, I suggest we make it clear like this:


```python
def get_unique_connection_parameters(connection_parameters):
    unique_by_cluster_name = {c.cluster_name: c for c in connection_parameters}
    return list(unique_by_cluster_name.values())
```

```python
        all_clusters = get_unique_connection_parameters(connection_parameters)
        privileged_clusters = get_unique_connection_parameters(
            (c for c in connection_parameters if (c.is_cluster_admin or cluster_admin))
        )

        threaded.run(
            self._init_oc_client,
            all_clusters,
            self._thread_pool_size,
            privileged=False,
        )
        threaded.run(
            self._init_oc_client,
            privileged_clusters,
            self._thread_pool_size,
            privileged=True,
        )
```

Reply:
Case 1: shouldnt be a problem - the cluster will be the same with `is_cluster_admin` and w/o, i.e., it doesn't matter which one wins (note, currently it is random anyways).

Case 2: the last one wins is fine here.

Basically this already happens in the original implementation mentioned [here](https://github.com/app-sre/qontract-reconcile/blob/ee0f52bf7f2551d00833d6858861ba16d1dd2357/reconcile/utils/oc.py#L1586-L1590). I am just porting that behavior, which is working fine for several months now in production - so Im very confident this will be ok.

Will add a function as you suggest to make it more clear - good idea!

Reply:
About the `is_cluster_admin` attribute: it is not considered in the rest of the code. This is the only part where it is used, i.e., to distinguish whether the cluster is initialized with `privileged` flag in the `threaded` call or w/o.

Reply:
I will highlight the intent through the variable naming instead then.

Reply:
I renamed the variables now and added a comment on why it is important to iterate over all parameters.

Reply:
`get_unique_connection_parameters` didn't change the execution order, should keep the same order as raw `connection_parameters`, `(c for c in connection_parameters if (c.is_cluster_admin or cluster_admin))` is iterator, so it didn't change order either.

Reply:
Ohhh you are right! I oversaw the extra brackets. Yeah that should work :)
Added your suggestion.

Reply:
I believe that ideally all this shoud be done in [oc_connection_parameters.py](https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/oc_connection_parameters.py). For now it will happen in `OCMap` to keep it as close to the original implementation as possible. Once the migration away from old `OC_Map` is finished, I will take a look to reduce the logic needed here.